posts/using-swagger-to-build-effective-apis.md

```diff
@@ -0,0 +1,382 @@
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Exercise'
+        '500':
+          description: Server error prevented successful retrieval
+  /exercises/{id}:
+    summary: Returns an exercise based on the ID passed
+    description: Returns an exercise based on the ID passed
+    get:
+      tags:
+        - Exercise ID
+      summary: "Returns an exercise based on the ID passed"
+      description: "Returns an exercise based on the ID passed"
+      parameters: 
+        - name: id
+          in: path
+          description: Exercise ID
+          required: true
+          schema:
+            type: integer
+            format: int64
+      operationId: 'getExerciseByID'
+      responses:
+        '200':
+          description: Successfully retrieved a single Exercise object
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Exercise'
+  /exercises/exercise:
+    summary: POST a new exercise
+    description: POST a new exercise
+    post:
+      tags:
+      - Exercise
+      operationId: 'postExercise'
+      summary: "POST a new exercise"
+      description: "POST a new exercise"
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Exercise'
+      responses:
+        '202':
+          description: Payload was accepted
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Exercise'
+        '400':
+          description: Back-end was unavailable
+```
+The example above shows two GET endpoints that define operations on the exercise resource defined in the components section of the specification.  You'll notice for each endpoint we define an HTTP Method (at least one), any relevant parameters associated with that method request, and the expected responses that come from that request path are mapped to components with the `$ref` tag.
+
+Pathing is a complex topic in API development.  There are many things to consider when defining API paths, like HTTP Headers, HTTP Methods, endpoint names, response types, versions, etc.  For example, you will notice each API path I have specified is a noun.  This noun is either singular or plural, depending on the number of objects returned by the response definition.  The only verbs in this API design are the HTTP verbs, which is as it should be.  Some developers may be tempted to say `http://www.myapi.com/getAllExercises`.  This is not good practice; behind the scenes this operations is a `GET getAllExercises`, which is redundant.  Instead, define your API paths as nouns and use the HTTP operations as verbs.  If you can concatenate the two and create a sensible phrase, you have successfully defined a logical endpoint like `POST exercise` or `GET exercises`.
```

Comment:
Agree. You could say that a `getAllExercises` endpoint would be RPC, not REST.

Reply:
Good point, modified this piece to add a nice article which breaks down the difference between REST and RPC